qemu/qemu.go

```diff
@@ -898,6 +901,94 @@ func (vhostuserDev VhostUserDevice) QemuParams(config *Config) []string {
 	return qemuParams
 }
 
+// PCIeRootPortDevice represents a memory balloon device.
+type PCIeRootPortDevice struct {
+	ID string // format: rp{n}, n>=0
+
+	Bus     string // default is pcie.0
+	Chassis string // (slot, chassis) pair is mandatory and must be unique for each pcie-root-port, >=0, default is 0
+	Slot    string // >=0, default is 0
+
+	Multifunction bool   // true => "on", false => "off", default is off
+	Addr          string // >=0, default is 0
+
+	// The PCIE-PCI bridge can be hot-plugged only into pcie-root-port that has 'bus-reserve' property value to
+	// provide secondary bus for the hot-plugged bridge.
+	BusReserve    string
+	Pref64Reserve string // reserve prefetched MMIO aperture, 64-bit
+	Pref32Reserve string // reserve prefetched MMIO aperture, 32-bit
+	MemReserve    string // reserve non-prefetched MMIO aperture, 32-bit *only*
+	IOReserve     string // IO reservation
+
+	ROMFile string // ROMFile specifies the ROM file being used for this device.
+}
+
+// QemuParams returns the qemu parameters built out of the PCIeRootPortDevice.
+func (b PCIeRootPortDevice) QemuParams(_ *Config) []string {
+	var qemuParams []string
+	var deviceParams []string
+	driver := PCIeRootPort
+
+	deviceParams = append(deviceParams, fmt.Sprintf("%s,id=%s", driver, b.ID))
+
+	if b.Bus == "" {
+		b.Bus = "pcie.0"
+	}
+	deviceParams = append(deviceParams, fmt.Sprintf("bus=%s", b.Bus))
+
+	if b.Chassis == "" {
+		b.Chassis = "0"
+	}
+	deviceParams = append(deviceParams, fmt.Sprintf("chassis=%s", b.Chassis))
+
+	if b.Slot == "" {
+		b.Slot = "0"
+	}
+	deviceParams = append(deviceParams, fmt.Sprintf("slot=%s", b.Slot))
+
+	multiifunction := "off"
+	if b.Multifunction {
+		multiifunction = "on"
+		if b.Addr == "" {
+			b.Addr = "0"
+		}
+		deviceParams = append(deviceParams, fmt.Sprintf("addr=%s", b.Addr))
+	}
+	deviceParams = append(deviceParams, fmt.Sprintf("multifunction=%v", multiifunction))
+
+	if b.BusReserve != "" {
+		deviceParams = append(deviceParams, fmt.Sprintf("bus-reserve=%s", b.BusReserve))
+	}
+
+	// the "pref32-reserve" and "pref64-reserve" hints are mutually exclusive.
+	if b.Pref64Reserve != "" {
```

Comment:
this means that `Pref64Reserve` takes precedence over `Pref32Reserve` ?  what happen if both are included? what happen if API user defines both? I think you should validate this in `(b PCIeRootPortDevice) Valid()`

Reply:
OK